### ISSUE TITLE
Implement asynchronous data retrieval via DICOMweb protocol

### ISSUE THREAD

**Issue by lassoan** [Slicer/Slicer#4743]
To allow Slicer to be launched from web browser (by clicking on a custom URL), we would need parallel, asynchronous data retrieval via DICOMweb (see related discussion in https://github.com/Slicer/Slicer/issues/3794).

**Comment by pieper** [comment]
I had started looking at this.  Best would be to improve the [python dicomweb-client](https://dicomweb-client.readthedocs.io/en/latest/introduction.html) code to use an [asyncio](https://docs.python.org/3/library/asyncio.html) [version of requests](https://pypi.org/project/aiohttp-requests/).

But to use that in Slicer, we need to tie asyncio in with Qt's event loop.  Closest I found was [asyncqt](https://github.com/gmarull/asyncqt), which would need to be ported from PyQt/PySide to work with PythonQt.  (Plus it looks messy - shouldn't Qt be able to abstract away the window vs unix differences?).

For our Slicer purposes it might be more straightforward to implement the async wado operations using native [Qt networking](https://doc.qt.io/qt-5/qtnetwork-programming.html).

**Comment by lassoan** [comment]
I would very much in favor of implementing this using Qt networking in CTK. Marco Nolden should be able to help with implementation, maintenance, and optimization, too.

**Comment by pieper** [comment]
Agreed.  There's not a lot of code required to implement the dicomweb part.  I think we could even imagine refactoring ctkDICOMDatabase to have either an sqlite or dicomweb backend. Or maybe some combination with sqlite or other local caching.

**Comment by lassoan** [comment]
The current database works nicely as a local cache, but the table views can be used to build queries - similarly to how it is done for DIMSE query/retrieve.

**Comment by pieper** [comment]
It would be good to have an interface to work with multiple dicom resources, local and remote.  The interface should be consistent so you don't need completely different codebases like you do now for dimse vs local cache.

A current use case for me is to iterate through subsets of tcia collections.  I download them via tcia downloader and add them to the slicer database then iterate and search using the slicer.dicomDatabase API, but it would be cleaner if I could get the with dicomweb or local with a single API.

**Comment by lassoan** [comment]
As I think about it more, the database may be usable as is. Only the background indexer would be need to be implementation specific (local filesystem/DIMSE/DICOMweb).

We could add a "database selector" list, which would contain "Local" and various remote databases. The would be all CTK DICOM database instances on local hard disk, but they would use different indexers. Remote databases would probably only query at patient level by default and only download a study/series when a patient or study is selected.

**Comment by jamesobutler** [comment]
@pieper Can you provide an update here, since it has been a couple years, regarding the state of asynchronous data retrieval via DICOMweb protocol? Is that specific protocol still the recommended one to use? Any other updates related to this topic to point to?

**Comment by pieper** [comment]
There's really no change on this.  The [python dicomweb_client](https://dicomweb-client.readthedocs.io/en/latest/introduction.html) code is still synchronous, so the app blocks and there are no overlapped network requests so performance is pretty bad.  There's still no pattern to integrate python asyncio with PythonQt, although I'm sure it could be done with some effort.  More generally I don't know of a dicomweb cli that works as well as browser-based access as in OHIF; if we did have a cli we could use it external to slicer and then load the results into the database and this would probably the best medium term solution.  It might be worth writing such a client using [javascript dicomweb_client](https://github.com/dcmjs-org/dicomweb-client) and node, but node would add a runtime dependency we don't want.  We could use the qSlicerWebWidget to access via javascript, seems like a major hack, but might realistically provide the best performance.
